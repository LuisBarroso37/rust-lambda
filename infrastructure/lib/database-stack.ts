import type { StackProps } from "aws-cdk-lib";
import { CfnOutput, RemovalPolicy, Stack } from "aws-cdk-lib";
import { AttributeType, BillingMode, Table } from "aws-cdk-lib/aws-dynamodb";
import type { Construct } from "constructs";

export class DatabaseStack extends Stack {
	constructor(scope: Construct, id: string, properties?: StackProps) {
		super(scope, id, properties);

		const table = new Table(this, "Table", {
			partitionKey: {
				name: "#p",
				type: AttributeType.STRING,
			},
			sortKey: {
				name: "#s",
				type: AttributeType.STRING,
			},
			billingMode: BillingMode.PAY_PER_REQUEST,
			pointInTimeRecovery: true,
			deletionProtection: true,
		});

		table.applyRemovalPolicy(RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE);

		new CfnOutput(this, "TableName", {
			description: "The name of the DynamoDB table",
			value: table.tableName,
			exportName: "Table-TableName",
		});

		new CfnOutput(this, "TableArn", {
			description: "The ARN of the DynamoDB table",
			value: table.tableArn,
			exportName: "Table-TableArn",
		});
	}
}
