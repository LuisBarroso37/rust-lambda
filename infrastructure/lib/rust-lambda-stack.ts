import * as path from "path";
import { CfnOutput, Fn, RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { RustFunction } from "cargo-lambda-cdk";
import { Architecture, FunctionUrlAuthType } from "aws-cdk-lib/aws-lambda";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";

export class RustLambdaStack extends Stack {
	constructor(scope: Construct, id: string, props?: StackProps) {
		super(scope, id, props);

		const tableArn = Fn.importValue("Table-TableArn");
		const tableName = Fn.importValue("Table-TableName");

		const environment = {
			TABLE_NAME: tableName,
		};

		const dynamodbReadPolicy = new PolicyStatement({
			actions: ["dynamodb:Query", "dynamodb:GetItem", "dynamodb:Scan"],
			resources: [tableArn, `${tableArn}/index/*`],
		});
		const dynamodbReadWritePolicy = new PolicyStatement({
			actions: [
				"dynamodb:Query",
				"dynamodb:GetItem",
				"dynamodb:Scan",
				"dynamodb:PutItem",
				"dynamodb:UpdateItem",
			],
			resources: [tableArn, `${tableArn}/index/*`],
		});

		const getAllFunction = new RustFunction(this, "GetAllFunction", {
			manifestPath: path.join(__dirname, "..", "..", "rust_lambda"),
			binaryName: "get_all",
			runtime: "provided.al2023",
			architecture: Architecture.ARM_64,
			environment,
		});

		getAllFunction.addToRolePolicy(dynamodbReadPolicy);

		new LogGroup(this, "LogGroup1", {
			logGroupName: `/aws/lambda/${getAllFunction.functionName}`,
			retention: RetentionDays.TEN_YEARS,
			removalPolicy: RemovalPolicy.RETAIN,
		});

		const insertFunction = new RustFunction(this, "InsertFunction", {
			manifestPath: path.join(__dirname, "..", "..", "rust_lambda"),
			binaryName: "insert",
			runtime: "provided.al2023",
			architecture: Architecture.ARM_64,
			environment,
		});

		insertFunction.addToRolePolicy(dynamodbReadWritePolicy);

		new LogGroup(this, "LogGroup2", {
			logGroupName: `/aws/lambda/${insertFunction.functionName}`,
			retention: RetentionDays.TEN_YEARS,
			removalPolicy: RemovalPolicy.RETAIN,
		});

		const getAllFunctionUrl = getAllFunction.addFunctionUrl({
			authType: FunctionUrlAuthType.NONE,
		});

		const insertFunctionUrl = insertFunction.addFunctionUrl({
			authType: FunctionUrlAuthType.NONE,
		});

		new CfnOutput(this, "GetAllFunctionUrl", {
			value: getAllFunctionUrl.url,
		});

		new CfnOutput(this, "InsertFunctionUrl", {
			value: insertFunctionUrl.url,
		});
	}
}
