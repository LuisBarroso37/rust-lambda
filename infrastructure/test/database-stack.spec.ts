import { expect, test } from "@jest/globals";
import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { DatabaseStack } from "../lib/database-stack";

test("Access Stack Snapshot Test", () => {
	const app = new App();
	const stack = new DatabaseStack(app, "DatabaseStack");
	const template = Template.fromStack(stack);

	expect(template.toJSON()).toMatchSnapshot();
});
