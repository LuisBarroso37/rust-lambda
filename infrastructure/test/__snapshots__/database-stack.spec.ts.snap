// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Stack Snapshot Test 1`] = `
{
  "Outputs": {
    "TableArn": {
      "Description": "The ARN of the DynamoDB table",
      "Export": {
        "Name": "Table-TableArn",
      },
      "Value": {
        "Fn::GetAtt": [
          "TableCD117FA1",
          "Arn",
        ],
      },
    },
    "TableName": {
      "Description": "The name of the DynamoDB table",
      "Export": {
        "Name": "Table-TableName",
      },
      "Value": {
        "Ref": "TableCD117FA1",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TableCD117FA1": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "#p",
            "AttributeType": "S",
          },
          {
            "AttributeName": "#s",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "#p",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "#s",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
